nova_vm/src/ecmascript/abstract_operations/operations_on_iterator_objects.rs

```diff
@@ -0,0 +1,302 @@
+    // 6. If innerResult.[[Type]] is throw, return ? innerResult.
+    // 7. If innerResult.[[Value]] is not an Object, throw a TypeError exception.
+    // 8. Return ? completion.
+    todo!()
+}
+
+/// [7.4.11 CreateIterResultObject ( value, done )](https://tc39.es/ecma262/#sec-createiterresultobject)
+///
+/// The abstract operation CreateIterResultObject takes arguments value (an
+/// ECMAScript language value) and done (a Boolean) and returns an Object that
+/// conforms to the IteratorResult interface. It creates an object that conforms
+/// to the IteratorResult interface.
+pub(crate) fn create_iter_result_object(agent: &mut Agent, value: Value, done: bool) -> Value {
+    // 1. Let obj be OrdinaryObjectCreate(%Object.prototype%).
+    // 2. Perform ! CreateDataPropertyOrThrow(obj, "value", value).
+    // 3. Perform ! CreateDataPropertyOrThrow(obj, "done", done).
+    // 4. Return obj.
+    todo!()
+}
+
+/// [7.4.12 CreateListIteratorRecord ( list )](https://tc39.es/ecma262/#sec-createlistiteratorRecord)
+///
+/// The abstract operation CreateListIteratorRecord takes argument list (a List
+/// of ECMAScript language values) and returns an Iterator Record. It creates
+/// an Iterator (27.1.1.2) object record whose next method returns the
+/// successive elements of list.
+pub(crate) fn create_list_iterator_record(agent: &mut Agent, list: &[Value]) -> JsResult<Value> {
+    // 1. Let closure be a new Abstract Closure with no parameters that captures list and performs the following steps when called:
+    // a. For each element E of list, do
+    // i. Perform ? GeneratorYield(CreateIterResultObject(E, false)).
+    // b. Return NormalCompletion(undefined).
+    // 2. Let iterator be CreateIteratorFromClosure(closure, empty, %IteratorPrototype%).
+    // 3. Return the Iterator Record { [[Iterator]]: iterator, [[NextMethod]]: %GeneratorFunction.prototype.prototype.next%, [[Done]]: false }.
+    todo!()
+}
+
+/// [7.4.13 IteratorToList ( iteratorRecord )](https://tc39.es/ecma262/#sec-iteratortolist)
+///
+/// The abstract operation IteratorToList takes argument iteratorRecord (an
+/// Iterator Record) and returns either a normal completion containing a List
+/// of ECMAScript language values or a throw completion.
+pub(crate) fn iterator_to_list(agent: &mut Agent, iterator_record: Object) -> JsResult<Vec<Value>> {
+    // 1. Let values be a new empty List.
+    let mut values = Vec::new();
+
+    // 2. Let next be true.
+    let mut next = Value::Boolean(true);
+
+    // 3. Repeat, while next is not false,
+    while next.is_true() {
+        // a. Set next to ? IteratorStep(iteratorRecord).
+        next = iterator_step(agent, iterator_record)?;
+
+        // b. If next is not false, then
+        if !next.is_false() {
+            // SAFETY: Because iterator_step returns either false or an object this is always safe.
+            let next = unsafe { next.try_into().unwrap_unchecked() };
```

Comment:
IMO it would be better if we changed `iterator_step`'s return type to `Option<Object>` and have the `false` state be `None`.

This would make this unsafe code obsolete and you would be able to do an `if let Some(next) = next {` above here.

Reply:
Sounds good!